File: python/src/cryptopals/set2/challenge11.py
from Crypto.Cipher import AES
import random
import secrets
from cryptopals.basics import padding


def encryption_oracle(content, mode_for_testing=None):
    prefix = secrets.token_bytes(random.randint(5, 10))  # NOSONAR
    suffix = secrets.token_bytes(random.randint(5, 10))  # NOSONAR
    padded = padding(prefix + content + suffix)
    key = secrets.token_bytes(16)
    if mode_for_testing is None:
        mode = AES.MODE_ECB if bool(random.getrandbits(1)) else AES.MODE_CBC  # NOSONAR
    else:
        mode = mode_for_testing
    if mode == AES.MODE_ECB:
        aes = AES.new(key, mode)
    elif mode == AES.MODE_CBC:
        iv = secrets.token_bytes(16)
        aes = AES.new(key, mode, iv)
    else:
        raise ValueError("Unsupported mode %s" % mode)
    return aes.encrypt(padded)


def encryption_detector(mode_for_testing=None):
    """
    >>> encryption_detector(AES.MODE_ECB)
    'ECB'
    >>> encryption_detector(AES.MODE_CBC)
    'CBC'
    """
    # Why 48?: Make sure that block1 and block2 are all zero
    content = b"\x00" * 48
    secret = encryption_oracle(content, mode_for_testing)
    block1 = secret[16:32]
    block2 = secret[32:48]
    return "ECB" if block1 == block2 else "CBC"

File: kotlin/src/main/kotlin/io/github/ocirne/cryptopals/set2/Challenge11.kt
package io.github.ocirne.cryptopals.set2

import io.github.ocirne.cryptopals.Basics.Extensions.repeat
import io.github.ocirne.cryptopals.Basics.padding
import io.github.ocirne.cryptopals.Oracle
import io.github.ocirne.cryptopals.crypto.AES
import kotlin.random.Random

class Challenge11Oracle(private val modeForTesting: String = "RANDOM"): Oracle {

    override fun encrypt(pt: ByteArray): ByteArray {
        val prefix = Random.nextBytes(Random.nextInt(5, 10))
        val suffix = Random.nextBytes(Random.nextInt(5, 10))
        val padded = padding(prefix + pt + suffix)
        val key = Random.nextBytes(16)
        val mode = if (modeForTesting == "RANDOM") {
            if (Random.nextBoolean()) "ECB" else "CBC"
        } else {
            modeForTesting
        }
        val aes = when (mode) {
            "ECB" ->
                AES.ECB(key)
            "CBC" -> {
                val iv = Random.nextBytes(16)
                AES.CBC(key, iv)
            }
            else -> throw IllegalStateException()
        }
        return aes.encrypt(padded)
    }
}

fun detectMode(oracle: Oracle): String {
    // Why 48?: Make sure that block1 and block2 are all zero
    val content = 0x00.toByte().repeat(48)
    val secret = oracle.encrypt(content)
    val block1 = secret.sliceArray(16..31)
    val block2 = secret.sliceArray(32..47)
    return if (block1.contentEquals(block2)) "ECB" else "CBC"
}
